### PR TITLE
MAINT: Throw PdfReadError if Trailer can't be read

### LINKED ISSUE #1279: TypeError: 'NoneType' object is not iterable
See https://github.com/py-pdf/PyPDF2/issues/1269 for further details.

## Environment

Which environment were you using when you encountered the problem?

```bash
$ python -m platform
Linux-5.4.0-122-generic-x86_64-with-glibc2.29

$ python -c "import PyPDF2;print(PyPDF2.__version__)"
2.10.3
```
## Code + PDF

This is a minimal, complete example that shows the issue:

```python
import PyPDF2
with open("pca_var.pdf", "rb") as f:
  pdfreader = PyPDF2.PdfFileReader(f, strict=False)
  metadata = pdfreader.metadata.copy() if pdfreader.metadata is not None else {}
```

PDF used above: [pca_var.pdf](https://github.com/tnd79bk/EbookMLCB/blob/2441571df00c59c57b431c74ef4000faaa795a52/Chapters/figs/07_DimemsionalityReduction/27_pca/latex/pca_var.pdf)

## Traceback

This is the complete Traceback I see:

```
Object 26 0 not defined.

Traceback (most recent call last):
  File "test.py", line 4, in <module>
    metadata = pdfreader.metadata.copy() if pdfreader.metadata is not None else {}
  File "/usr/local/lib/python3.8/dist-packages/PyPDF2/_reader.py", line 329, in metadata
    retval.update(obj)  # type: ignore
TypeError: 'NoneType' object is not iterable
```

### PR DESCRIPTION
**Pull request by ediamondscience**
… Added myself to contributor list.

In reference to #1279 I've modified _reader.py to throw a PdfReadError if the trailer doesn't exist, can't be read, or does not point to the object containing the metadata for the document. 

Three pdf files were uploaded in relation to aforementioned issue, pca_var.pdf contained no trailer, and the other two had corrupted declarations that PyPDF2 could not read. I'll include them here for those interested. 

From Rotary Potentiometer.pdf:
`trailer^M
<</Size 267/Prev 2815238/Root 176 0 R/Info 174 0 R/ID[<8CC42D96833D2A438E62CB60D00F610F><5C480BCE388CB6408C83C951FDD0DA22>]>>^M
startxref^M
0^M
%%EOF^M`

From Status_v1_Reviewers-Guide.pdf:
`trailer`
`/Size`
`447`
`/Root`
`4`
`0`
`R`
`/Info`
`5`
`0`
`R`
`startxref`
`6128518`
(sorry for using multiple code blocks, but I think it's important to show that there are line breaks for each of these lines in my terminal)

We should probably figure out why these trailers are displayed like this. If I had to guess it might be because of a different encoding for line breaks. If we can update the way that we read the trailer to account for this we might be able to read useful information out of more files. 

All three files from the issue should now throw an error when the test code is run. This should close #1279

### CODE REVIEW

**Comment by MartinThoma** [comment]
Can you please add a test that covers this? You mentioned that https://github.com/p-sampson-II/SEEDLab-TAs-2022/blob/378e6f583ed9de274a8bda9240dfa7aa6872a733/Alex%20Curtis/Datasheets/Rotary%20Potentiometer.pdf shows the issue. You can use the file via:

```python
url = "https://github.com/p-sampson-II/SEEDLab-TAs-2022/blob/378e6f583ed9de274a8bda9240dfa7aa6872a733/Alex%20Curtis/Datasheets/Rotary%20Potentiometer.pdf"
name = "github-Rotary-Potentiometer.pdf"
reader = PdfReader(BytesIO(get_pdf_from_url(url, name=name)))
```

File: PyPDF2/_reader.py [hunk 1]
```diff
@@ -326,8 +326,11 @@ def metadata(self) -> Optional[DocumentInformation]:
             return None
         obj = self.trailer[TK.INFO]
         retval = DocumentInformation()
-        retval.update(obj)  # type: ignore
-        return retval
+        if isinstance(obj, type(None)):
+            raise PdfReadError("trailer not found or does not point to document information directory")
+        else:
+            retval.update(obj)  # type: ignore
+            return retval
```
Comment (MartinThoma) [inline PyPDF2/_reader.py:333]:
```suggestion
        retval.update(obj)  # type: ignore
        return retval
```

**Comment by MartinThoma** [comment]
Thank you for your work on this one!

I've just noticed that you've added `resources/unreadablemetadata.pdf` which has 2MB. In order to keep this repository small (and thus easy to clone, also with bad connections), I would like this file to be in https://github.com/py-pdf/sample-files/ .

Could you please add the file there? (I can also do it + use the "Contributed-by" feature of Github to attribute it to you)

**Comment by MartinThoma** [comment]
Thank you for your work on this one!

I've just noticed that you've added `resources/unreadablemetadata.pdf` which has 2MB. In order to keep this repository small (and thus easy to clone, also with bad connections), I would like this file to be in https://github.com/py-pdf/sample-files/ .

Could you please add the file there? (I can also do it + use the "Contributed-by" feature of Github to attribute it to you)

File: PyPDF2/_reader.py (hunk 1 above)
Comment (ediamondscience) [reply PyPDF2/_reader.py]: I'm curious about this one. Why remove the else statement? Is it for performance or something else? Should I have a try statement in place of the if?

**Comment by ediamondscience** [comment]
@MartinThoma I've added a pull request to get that file into the correct repository and updated the file location in test_reader.py 

The commit may fail some tests in this repository before the sample files are merged, but I've tested it on my own rig here with pytest and it should work perfectly after the merge.

**Comment by MartinThoma** [comment]
I was just blocked at work and added the updated submodule to PyPDF2 :-)

**Comment by MartinThoma** [comment]
Could you please do `git rm resources/unreadablemetadata.pdf`?

**Comment by ediamondscience** [comment]
Looks like the the reader causes some issues if the metadata is unreadable. Is it the right move to throw a PdfReadError for this situation?

**Comment by MartinThoma** [comment]
Sorry, my mistake. I'll fix the test tomorrow 🙈

**Pull request merged**